File: qt/Url.cpp
//
// Created by XLX on 2026/5/12.
//

#include "Url.h"

#include "util.h"

Url::Url(const QString &url, const QString &sni, QObject *parent) : Url(url, parent) {
    bindings::Url_set_sni(this->raw_ptr, sni.toUtf8());
}

Url::Url(const QString &url, QObject *parent) : QObject(parent) {
    char *err = nullptr;
    this->raw_ptr = bindings::Url_new(url.toUtf8(), &err);
    util::check_err(err);
}

void Url::addParam(const QString &name, const QString &value) const {
    char *err = bindings::Url_add_param(this->raw_ptr, name.toUtf8(), value.toUtf8());
    util::check_err(err);
}

void Url::removeParam(const QString &name) const {
    char *err = bindings::Url_remove_param(this->raw_ptr, name.toUtf8());
    util::check_err(err);
}

bindings::Url *Url::take() {
    bindings::Url *raw = this->raw_ptr;
    this->raw_ptr = nullptr;
    return raw;
}

Url::~Url() {
    if (this->raw_ptr == nullptr)return;
    bindings::Url_drop(this->raw_ptr);
    this->raw_ptr = nullptr;
}

File: java/src/main/java/org/xllgl2017/Url.java
package org.xllgl2017;

import com.sun.jna.Pointer;
import com.sun.jna.ptr.PointerByReference;

import java.util.HashMap;

import static org.xllgl2017.ReqrioLibrary.REQRIO;

public class Url implements AutoCloseable {
    private Pointer raw;

    public Pointer pointer() throws Exception {
        if (this.raw == null) throw new Exception("Url had dropped");
        return this.raw;
    }

    /// 初始化
    ///
    /// @param url :请求地址
    public Url(String url) throws Exception {
        PointerByReference err = new PointerByReference();
        this.raw = REQRIO.Url_new(url, err);
        util.check_err_pointer(err);
    }

    /// @param sni :域名
    public Url(String url, String sni) throws Exception {
        this(url);
        this.setSni(sni);
    }

    /// @param params url请求参数，value应为未编码
    public Url(String url_str, HashMap<String, String> params) throws Exception {
        this(url_str);
        for (String key : params.keySet()) {
            this.addParam(key, params.get(key));
        }
    }

    /// 添加一个请求参数，若这个值已存在则会被覆盖
    ///
    /// @param name  :参数名
    /// @param value :参数值
    public void addParam(String name, String value) throws Exception {
        try {
            util.check_err(REQRIO.Url_add_param(this.pointer(), name, value));
        } catch (Exception e) {
            close();
            throw e;
        }
    }

    /// 删除一个参数
    ///
    /// @param name :待删除的参数名
    public void removeParam(String name) throws Exception {
        try {
            util.check_err(REQRIO.Url_remove_param(this.pointer(), name));
        } catch (Exception e) {
            close();
            throw e;
        }
    }

    public void setRaw(Pointer raw) {
        this.raw = raw;
    }

    /// @param sni :域名，在使用ip url时设置
    public void setSni(String sni) throws Exception {
        try {
            util.check_err(REQRIO.Url_set_sni(this.pointer(), sni));
        } catch (Exception e) {
            close();
            throw e;
        }
    }

    @Override
    public void close() {
        if (this.raw == null) return;
        REQRIO.Url_drop(this.raw);
        this.raw = null;
    }
}
